acs-edge/lib/devices/driver.ts

```diff
@@ -0,0 +1,210 @@
+            id:         this.id,
+            msg:        "active",
+            payload:    Buffer.from("ONLINE"),
+        });
+        /* We do not emit "open" here, we wait for the negotiation with
+         * the driver. */
+    }
+
+    close () {
+        this.broker.publish({
+            id:         this.id,
+            msg:        "active",
+            payload:    Buffer.from("OFFLINE"),
+        });
+        this.broker.off("message", this.#message.bind(this));
+        this.emit("close");
+    }
+
+    
+
+    readMetrics(metrics: Metrics, payloadFormat?: string, delimiter?: string) {
+        const poll = metrics.addresses
+            .filter(a => this.topics.has(a))
+            .map(a => this.topics.get(a))
+            .join("\n");
+        this.broker.publish({
+            id:         this.id,
+            msg:        "poll",
+            payload:    Buffer.from(poll),
+        });
+    }
+
+    writeMetrics(metrics: Metrics, writeCallback: Function, payloadFormat: serialisationType, delimiter?: string) {
+        let err = null;
+        // Do whatever connection specific stuff you need to in order to
+        // write to the device
+
+        // Call the writeCallback when complete, setting the error if
+        // necessary
+        writeCallback(err);
+    }
+
+    /**
+     *
+     * @param metrics Metrics object to watch
+     * @param payloadFormat String denoting the format of the payload
+     * @param delimiter String specifying the delimiter character if needed
+     * @param interval Time interval between metric reads, in ms
+     * @param deviceId The device ID whose metrics are to be watched
+     * @param subscriptionStartCallback A function to call once the subscription has been setup
+     */
+    async startSubscription(metrics: Metrics, payloadFormat: serialisationType, delimiter: string, interval: number, deviceId: string, subscriptionStartCallback: Function) {
+        const addrs = metrics.addresses;
+        const unassigned = addrs.filter(a => !this.topics.has(a));
+        for (const a of unassigned) {
+            const dt = this.#newTopic();
+            this.addrs.set(dt, a);
+            this.topics.set(a, dt);
+        }
+        const topics = new Set(addrs.map(a => this.topics.get(a)!))
+        this.groups.set(deviceId, {
+            poll:   interval,
+            addrs:  topics,
+        });
+
+        if (this.status != "DOWN")
+            this.#send_addrs();
+
+        super.startSubscription(metrics, payloadFormat, delimiter, interval,
+            deviceId, subscriptionStartCallback);
+    }
+
+    /**
+     * Stop a previously registered subscription for  metric changes.
+     * @param deviceId The device ID we are cancelling the subscription for
+     * @param stopSubCallback A function to call once the subscription has been cancelled
+     */
+    //async stopSubscription(deviceId: string, stopSubCallback: Function) {
+    //}
+    
+    #newTopic () {
+        while (true) {
```

Comment:
What does this do, and why?

Reply:
The edge driver protocol maps addresses (which might be long strings and unsuitable for MQTT topic components) into data topic names assigned by the edge agent. This generates a new random number to use as a data topic name, and checks it actually is new before returning it.